keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+can hang forever waiting for a response from the server. This is certainly true for
+HTTP/1x where [net/http does not execute the request handler in a new goroutine](https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/server.go#L2031-L2039).
+
+> It's worth noting that by default `client-go` does not enforce any timeout globally.
+
+With the removal of the `timeout` filter, we need to ensure that a request handler that
+blocks on the server should never block a client that does not enforce timeout on its side.
+
+In fact, we can reproduce a "forever" hanging client with http/1x:
+```
+func TestClientHangingForeverWithHTTP1(t *testing.T) {
+	clientDoneCh, handlerDoneCh := make(chan struct{}), make(chan error, 1)
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		defer close(handlerDoneCh)
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(200 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetWriteDeadline, but got: %v", err)
+			return
+		}
+		if err := ctrl.SetReadDeadline(time.Now().Add(100 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
+			return
+		}
+
+		<-clientDoneCh
+	}))
+
+
+	defer server.Close()
+	server.StartTLS()
+
+	client := server.Client()
+	func() {
+		defer close(clientDoneCh)
+		client.Get(server.URL + "/foo")
+	}()
+
+	select {
+	case <-handlerDoneCh:
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("expected the request handler to have terminated")
+	}
+}
+```
+Even though a read/write deadline is set, the client never receives any 
+response or error from the server. We have opened an [issue](https://github.com/golang/go/issues/65526)
+with the golang team to track this.
+
+Mitigation Steps:
+- a) If this is a bug, we can wait until the fix is in the net/http package.
+- b) Otherwise, we can enable this feature only for http/2.0 request(s), all
+  http/1x request(s) will continue to be served using the `Timeout` filter.
+
+We also need to add unit tests to verify that a blocked request handler
+does not block the corresponding client consequently:
+```
+neverDoneCh := make(chan struct{})
+func(w http.ResponseWriter, req *http.Request) {
+		<-neverDoneCh
+	}))
+```
+
+
+##### Request Handler Blocking Indefinitely:
+
+Once read/write deadline exceeds, the corresponding request handler goroutine will
+finish its run by returning an error, or throwing a panic as soon as it uses the
+`ResponseWriter` object by invoking `Write` or `FlushError`
+
+This is in par with the `timeout` filter.
+
+
+##### Impact When an HTTP/1x Connection is Hijacked
+
+The hijacker is responsible for managing the state of the hijacked
+connection. Add a test to demonstrate that setting a read/write deadline before
+the connection is hijacked does not interfere with the operation of the hijacker.
```

Comment:
Will we do this by finding all our current hijack type assertions, figure out what they're used for, and ensure we have at least integration tests (we get a real handler for these) to be sure they work?

Reply:
I think a new unit test that verifies that read/write deadline are reset and the hijacked connection can live beyond the original read/write deadline should suffice, thoughts?

Reply:
> I think a new unit test that verifies that read/write deadline are reset and the hijacked connection can live beyond the original read/write deadline should suffice, thoughts?

I'm less confident. I'm worried about what trip dependencies lie within the code around the particulars of our usage, not about the hijack in general.  How many instances do we have?  Looks eight `grep -r "(http.Hijack" pkg staging/ | grep -v "_test"`?

Reply:
only two instances invoke `Hijack`, the rest are the `ResponseWriter` delegators, `apimachinery/pkg/util/proxy/upgradeaware.go` and `apimachinery/pkg/util/httpstream/spdy/upgrade.go`.  We can add unit tests around these to verify no change in behavior, if it is possible to write integration tests, we can go for that as well.

Reply:
I added these in the integration test section